### DISCUSSION TITLE
Python Wrapper - Possible memory leak using PIL

### DISCUSSION THREAD

**Discussion by gregtws** [zxing-cpp/zxing-cpp#762] [Q&A]
When loading a large image (~6 mbytes; ~10k x 10k pixels RGB), I'm noticing a memory leak between repeated loads. 

I guess I'm looking to ideas on tracing the source of the leak; I suspect its an issue with PIL or maybe how zxing uses PIL (the main allocation is via PILs' `Image.tobytes` which is triggered when we process the image like an array).

It seems like PIL isn't freeing up the memory, but I can't figure out why.  The net result is a leak of ~300mbytes per load.   I don't see this behavior with Openvc / `cv2.imread()`.

The flow is equivalent to calling the following repeatedly:

```
img = Image.open("file.png")
decoded = zxingcpp.read_barcodes(img, formats=zxingcpp.BarcodeFormat.MatrixCodes, return_errors=True)
img.close() # also tried deleting and manually calling a gc.collect()
```

Any ideas?

Using PIL 10.3.0, zxing 2.2.0 release.

**Comment by axxel** [comment]
My first thought: maybe this is a bug in the `zxing-cpp` python module and it might be depending on the pixel format of the image data. If you are on Linux, can you tell me the output of `file file.png`? I'm looking for something like `PNG image data, 2640 x 2064, 8-bit/color RGBA, non-interlaced`.

Second thought: can you try if the following makes any difference?
```python
with Image.open("file.png") as img:
    decoded = zxingcpp.read_barcodes(img, formats=zxingcpp.BarcodeFormat.MatrixCodes, return_errors=True)
```

**Reply by gregtws** [reply]
Thanks for taking a look.   Looks like the troublesome image is a jpeg not png, file output is:

```
JPEG image data, JFIF standard 1.01, aspect ratio, density 300x300, segment length 16, baseline, precision 8, 6000x19367, components 3
```

(Pillow sees this as an RGB file)

I tried wrapping a context around it, no change.  Also tried without luck:

* Setting (one at a time): `is_pure=True`, `try_rotate=False`, `try_downscale=False`.
* Loading as a PNG

Given allocations are coming from `tobytes` it might be how we traverse the buffer (__array_interface__ access triggers reading the file and creating the bytes result in `data`, but I don't understand why that wouldn't be freed.   (guessing memory views wouldn't help here either, but I suppose it could).

**Reply by gregtws** [reply]
It could be related to python-pillow/Pillow/issues/7935 but in my test I'm not copying an image (and the `PILLOW_BLOCKS_MAX` setting doesn't change behavior)

**Reply by axxel** [reply]
Can you test if the code from [this comment](https://github.com/python-pillow/Pillow/issues/5797#issuecomment-954655638) triggers your memory leak? It sounds pretty much exactly like our use case but should have been fixed a long time ago already.

Following this discussion led me to [this blog post](https://uploadcare.com/blog/fast-import-of-pillow-images-to-numpy-opencv-arrays/) which could be used as a base to potentially fix the issue at hand and make the code a little faster. But I don't have time right now to spend on this.

**Reply by gregtws** [reply]
Using `numpy.array` or `asarray` still triggers the leak.

Good find on the blog post, it does seem to minimize the leak (maybe 100mb the first time and a few mb each subsequent call).  So definitely better than 300mb per call, still more leaky than using opencv directly.

I'm not sure I'll have time to dig into a fix further at the moment, if I do or find more from PIL I'll send along a note/MR.

Thanks again!
